app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CryptoCalculatorImpl.kt

```diff
@@ -0,0 +1,111 @@
+    private var decimalClicked: Boolean = false
+    private var decimalCounter: Int = 0
+    private var taskBuilder: BackgroundCryptoTaskBuilder = BackgroundCryptoTaskBuilder(calculator, this)
+    private var testing = false
+
+    init {
+        setValue("0.000000")
+        handleClear()
+    }
+
+    fun numpadClicked(id: Int) {
+        when (id) {
+            R.id.btn_decimal -> decimalClick()
+            R.id.btn_0 -> addDigit(0)
+            R.id.btn_1 -> addDigit(1)
+            R.id.btn_2 -> addDigit(2)
+            R.id.btn_3 -> addDigit(3)
+            R.id.btn_4 -> addDigit(4)
+            R.id.btn_5 -> addDigit(5)
+            R.id.btn_6 -> addDigit(6)
+            R.id.btn_7 -> addDigit(7)
+            R.id.btn_8 -> addDigit(8)
+            R.id.btn_9 -> addDigit(9)
+        }
+    }
+
+    fun addDigit(i: Int) {
+        if ((number != "" || i != 0) && decimalCounter < 5) {
+            number = number + i
+            if (decimalClicked) {
+                decimalCounter++
+            }
+            setValue(String.format("%,.6f", number.toDouble()))
+        }
+    }
+
+    fun decimalClick() {
+        if (number == "") {
+            number = "0."
+        } else if (!decimalClicked) {
+            number += "."
+        }
+        decimalClicked = true
+    }
+
+    fun handleDelete() {
+        if (number.length <= 1) {
+            return handleClear()
+        } else if (decimalClicked && decimalCounter<5) {
+            number = number.dropLast(1 + decimalCounter)
+            decimalClicked = false
+            decimalCounter = 0
+        } else if (decimalClicked) {
+            decimalCounter--
+            number = number.dropLast(1)
+        } else {
+            number = number.dropLast(1)
+        }
+        setValue(String.format("%,.6f", number.toDouble()))
+    }
+
+    fun handleClear() {
+        number = ""
+        decimalClicked = false
+        decimalCounter = 0
+        setValue("0.000000")
+    }
+
+    // will be used by the 3 money action buttons to display the result
+    fun overwriteNumber(newNumber: Double) {
+        number = ""
+        decimalClicked = false
+        decimalCounter = 0
+        setValue(String.format("%,.6f", newNumber))
+    }
+
+    private fun setValue(value: String) {
+        mCallback!!.setValue(value)
+    }
+
+    private fun getResult() = Formatter.stringToDouble(mCallback!!.getResult())
+
+    fun calculateCrypto2(from: String, to: String) {
+        var response: AsyncTask<Void, Void, String>? = taskBuilder.from(from).to(to).build().execute()
+    }
+
+    fun supersedeBuilder(builder: BackgroundCryptoTaskBuilder) {
+        taskBuilder = builder
+        testing = true
+    }
+
+    fun calculateCrypto(cryptoFROM: String, cryptoTO: String, cryptoActivity: CryptoActivity) {
+        if (testing) {
+            calculateCrypto2(cryptoFROM, cryptoTO)
+        } else {
+            var response = GetCryptoTask(cryptoFROM, cryptoTO, cryptoActivity, this).execute()
```

Comment:
I think you should be using the builder here to return a new GetCryptoTask or a FakeCryptoTask. Using polymorphism rather than this testing flag.

Reply:
Contact me for more info xD

Reply:
Indeed this could be refactored to be cleaner. I shall implement it.